File: src/engine/sandhi.rs
use fancy_regex::Regex;
use lazy_static::lazy_static;
use crate::engine::declension::apply_natva;

struct UpasargaRule {
    pattern: Regex,
    canonical: &'static str,
    prepend: &'static str,
}

lazy_static! {
    static ref UPASARGA_PATTERNS: Vec<UpasargaRule> = vec![
        UpasargaRule { pattern: Regex::new(r"^samudA").unwrap(), canonical: "sam + ud + AN", prepend: "" },
        UpasargaRule { pattern: Regex::new(r"^sa[mMnYNR]u[dtcjNYRnl]A").unwrap(), canonical: "sam + ud + AN", prepend: "" },
        UpasargaRule { pattern: Regex::new(r"^sa[mMnYNR]u[dtcjNYRnl]").unwrap(), canonical: "sam + ud", prepend: "" },
        UpasargaRule { pattern: Regex::new(r"^vyA").unwrap(), canonical: "vi + AN", prepend: "" },
        UpasargaRule { pattern: Regex::new(r"^pratyA").unwrap(), canonical: "prati + AN", prepend: "" },
        UpasargaRule { pattern: Regex::new(r"^vy").unwrap(), canonical: "vi", prepend: "" },
        UpasargaRule { pattern: Regex::new(r"^vi").unwrap(), canonical: "vi", prepend: "" },
        UpasargaRule { pattern: Regex::new(r"^praty").unwrap(), canonical: "prati", prepend: "" },
        UpasargaRule { pattern: Regex::new(r"^prati").unwrap(), canonical: "prati", prepend: "" },
        UpasargaRule { pattern: Regex::new(r"^sa[mMnYNR]").unwrap(), canonical: "sam", prepend: "" },
        UpasargaRule { pattern: Regex::new(r"^sam").unwrap(), canonical: "sam", prepend: "" },
        UpasargaRule { pattern: Regex::new(r"^pra").unwrap(), canonical: "pra", prepend: "" },
        UpasargaRule { pattern: Regex::new(r"^A").unwrap(), canonical: "AN", prepend: "" },
    ];
}

pub fn get_upasarga_splits(word: &str) -> Vec<(String, String)> {
    let mut splits = Vec::new();
    for rule in UPASARGA_PATTERNS.iter() {
        if let Ok(Some(mat)) = rule.pattern.find(word) {
            let end_idx = mat.end();
            let stripped = format!("{}{}", rule.prepend, &word[end_idx..]);
            if stripped.chars().count() >= 2 {
                splits.push((rule.canonical.to_string(), stripped));
            }
        }
    }
    splits
}

pub fn apply_upasarga_sandhi(upasarga_str: &str, form: &str) -> String {
    if upasarga_str.is_empty() { return form.to_string(); }
    let prefixes: Vec<&str> = upasarga_str.split('+').map(|s| s.trim()).collect();
    let mut result = form.to_string();
    
    for mut p in prefixes.into_iter().rev() {
        if p == "AN" { p = "A"; }
        
        let p_len = p.len();
        let p_last = p.chars().last().unwrap_or(' ');
        let r_first = result.chars().next().unwrap_or(' ');
        let is_vowel = |c: char| "aAiIuUfFeEoO".contains(c);
        
        if p_last == 'i' && is_vowel(r_first) {
            result = format!("{}y{}", &p[..p_len-1], result);
        } else if p_last == 'u' && is_vowel(r_first) {
            result = format!("{}v{}", &p[..p_len-1], result);
        } else if p_last == 'a' || p_last == 'A' {
            if r_first == 'a' || r_first == 'A' { result = format!("{}A{}", &p[..p_len-1], &result[1..]); }
            else if r_first == 'i' || r_first == 'I' { result = format!("{}e{}", &p[..p_len-1], &result[1..]); }
            else if r_first == 'u' || r_first == 'U' { result = format!("{}o{}", &p[..p_len-1], &result[1..]); }
            else if r_first == 'f' || r_first == 'F' { result = format!("{}ar{}", &p[..p_len-1], &result[1..]); }
            else { result = format!("{}{}", p, result); }
        } else if p_last == 'm' {
            if !is_vowel(r_first) { result = format!("{}M{}", &p[..p_len-1], result); }
            else { result = format!("{}{}", p, result); }
        } else if p_last == 'd' {
            if "kKqQpPzSs".contains(r_first) { result = format!("{}t{}", &p[..p_len-1], result); }
            else if "cC".contains(r_first) { result = format!("{}c{}", &p[..p_len-1], result); }
            else if "jJ".contains(r_first) { result = format!("{}j{}", &p[..p_len-1], result); }
            else { result = format!("{}{}", p, result); }
        } else {
            result = format!("{}{}", p, result);
        }
        
        result = apply_natva(&result);
    }
    result
}

File: engine/sandhi.go
package engine

import (
	"regexp"
	"strings"
	"unicode/utf8"
)

type upasargaRule struct {
	pattern   *regexp.Regexp
	canonical string
	prepend   string
}

var upasargaPatterns []upasargaRule

func init() {
	upasargaPatterns = []upasargaRule{
		{pattern: regexp.MustCompile(`^samudA`), canonical: "sam + ud + AN", prepend: ""},
		{pattern: regexp.MustCompile(`^sa[mMnYNR]u[dtcjNYRnl]A`), canonical: "sam + ud + AN", prepend: ""},
		{pattern: regexp.MustCompile(`^sa[mMnYNR]u[dtcjNYRnl]`), canonical: "sam + ud", prepend: ""},
		{pattern: regexp.MustCompile(`^vyA`), canonical: "vi + AN", prepend: ""},
		{pattern: regexp.MustCompile(`^pratyA`), canonical: "prati + AN", prepend: ""},
		{pattern: regexp.MustCompile(`^vy`), canonical: "vi", prepend: ""},
		{pattern: regexp.MustCompile(`^vi`), canonical: "vi", prepend: ""},
		{pattern: regexp.MustCompile(`^praty`), canonical: "prati", prepend: ""},
		{pattern: regexp.MustCompile(`^prati`), canonical: "prati", prepend: ""},
		{pattern: regexp.MustCompile(`^sa[mMnYNR]`), canonical: "sam", prepend: ""},
		{pattern: regexp.MustCompile(`^sam`), canonical: "sam", prepend: ""},
		{pattern: regexp.MustCompile(`^pra`), canonical: "pra", prepend: ""},
		{pattern: regexp.MustCompile(`^A`), canonical: "AN", prepend: ""},
	}
}

// GetUpasargaSplits returns potential upasarga splits as pairs of [canonical, stripped_stem]
func GetUpasargaSplits(word string) [][2]string {
	var splits [][2]string
	for _, r := range upasargaPatterns {
		loc := r.pattern.FindStringIndex(word)
		if loc != nil {
			endIdx := loc[1]
			stripped := r.prepend + word[endIdx:]
			if utf8.RuneCountInString(stripped) >= 2 {
				splits = append(splits, [2]string{r.canonical, stripped})
			}
		}
	}
	return splits
}

// ApplyUpasargaSandhi applies sandhi rules when combining prefixes with verb forms
func ApplyUpasargaSandhi(upasargaStr string, form string) string {
	if upasargaStr == "" || form == "" {
		return form
	}

	rawPrefixes := strings.Split(upasargaStr, "+")
	prefixes := make([]string, 0, len(rawPrefixes))
	for _, s := range rawPrefixes {
		p := strings.TrimSpace(s)
		if p != "" {
			prefixes = append(prefixes, p)
		}
	}

	result := form

	for i := len(prefixes) - 1; i >= 0; i-- {
		p := prefixes[i]
		if p == "AN" {
			p = "A"
		}

		pRunes := []rune(p)
		if len(pRunes) == 0 {
			continue
		}

		pLast := pRunes[len(pRunes)-1]
		pMinusOne := string(pRunes[:len(pRunes)-1])

		rRunes := []rune(result)
		if len(rRunes) == 0 {
			result = p + result
			continue
		}

		rFirst := rRunes[0]
		rMinusFirst := string(rRunes[1:])

		isVowel := func(c rune) bool {
			return strings.ContainsRune("aAiIuUfFxXeEoO", c)
		}

		if pLast == 'i' || pLast == 'I' {
			if rFirst == 'i' || rFirst == 'I' {
				result = pMinusOne + "I" + rMinusFirst
			} else if isVowel(rFirst) {
				result = pMinusOne + "y" + result
			} else {
				result = p + result
			}
		} else if pLast == 'u' || pLast == 'U' {
			if rFirst == 'u' || rFirst == 'U' {
				result = pMinusOne + "U" + rMinusFirst
			} else if isVowel(rFirst) {
				result = pMinusOne + "v" + result
			} else {
				result = p + result
			}
		} else if pLast == 'a' || pLast == 'A' {
			if rFirst == 'a' || rFirst == 'A' {
				result = pMinusOne + "A" + rMinusFirst
			} else if rFirst == 'i' || rFirst == 'I' {
				result = pMinusOne + "e" + rMinusFirst
			} else if rFirst == 'u' || rFirst == 'U' {
				result = pMinusOne + "o" + rMinusFirst
			} else if rFirst == 'f' || rFirst == 'F' {
				result = pMinusOne + "ar" + rMinusFirst
			} else if rFirst == 'e' || rFirst == 'E' {
				result = pMinusOne + "E" + rMinusFirst
			} else if rFirst == 'o' || rFirst == 'O' {
				result = pMinusOne + "O" + rMinusFirst
			} else {
				result = p + result
			}
		} else if pLast == 'm' {
			if !isVowel(rFirst) {
				result = pMinusOne + "M" + result
			} else {
				result = p + result
			}
		} else if pLast == 'd' {
			if strings.ContainsRune("kKqQpPzSstT", rFirst) {
				result = pMinusOne + "t" + result
			} else if strings.ContainsRune("cC", rFirst) {
				result = pMinusOne + "c" + result
			} else if strings.ContainsRune("jJ", rFirst) {
				result = pMinusOne + "j" + result
			} else if rFirst == 'l' {
				result = pMinusOne + "l" + result
			} else if rFirst == 'n' || rFirst == 'm' {
				result = pMinusOne + "n" + result
			} else {
				result = p + result
			}
		} else if pLast == 'r' {
			if rFirst == 'r' {
				if len(pRunes) >= 2 && pRunes[len(pRunes)-2] == 'i' {
					pMinusTwo := string(pRunes[:len(pRunes)-2])
					result = pMinusTwo + "I" + result
				} else if len(pRunes) >= 2 && pRunes[len(pRunes)-2] == 'u' {
					pMinusTwo := string(pRunes[:len(pRunes)-2])
					result = pMinusTwo + "U" + result
				} else {
					result = p + result
				}
			} else {
				result = p + result
			}
		} else {
			result = p + result
		}

		result = applyNatva(result)
	}

	return result
}